Re-ARM - hardware SPI with Viki & RRD full graphics support

Bob-the-Kuhn opened the pull request:
1) Created hardware SPI - this just implements the same commands as in the software SPI.  No token or channel support.

2) Modified VIKI & RRD Full Graphics drivers to work with it.  I don't know if the RRD Smart Controller works with this.

3) Modified HAL.h so that the SPI speeds match the definitions in the SD card software.

4) disabled software SPI

5) added ENV for J-Link debug probe.

6) set SPI speed for RRD Full Graphics to 1 MHz independent of SD card SPI speed

----

@p3p - let me know if this is part of what you envisioned for the hardware SPI.  I'll be happy to add/delete/correct as you wish.

----

Somewhere along the line I lost my working (well, at least it compiled) copy of the general LCD implementation for ARM.  Hopefully I'll have that available for review in less than a week.

p3p committed:
HAL for Re:ARM (LPC1768) architecture

p3p committed:
Implement HAL and apply macros across code-base
Implement AVR Platform

teemuatlut committed:
HAL for 32-bit Teensy (3.5, 3.6) architecture

p3p committed:
Apply Travis update workaround while a permanent solution is found

p3p committed:
HAL for DUE architecture

p3p committed:
PlatformIO Travis test

spi.h renamed to avoid conflicts with base libraries

p3p committed:
Initial HAL SPI API

p3p committed:
Base HAL SPI Changes

p3p committed:
Debug Pins Support

thinkyhead committed:
Marlin bugfix-2.0.x Version/README

Bob-the-Kuhn committed:
minimize SD card reads when using SDCARD_SORT_ALPHA (Re-ARM board)

Because of the Re-ARM card's pinout there is only one SPI connected to
the RepRap Discount Full Graphic LCD display.  The LCD responds to ANY
SCK transitions no matter if it's enable is inactive.  The result is
garbage (usually bars) on the LCD display whenever there is SD card
activity.

This code minimizes this by only accessing the SD card when changing
directory levels if :
SDCARD_SORT_ALPHA is enabled
SDSORT_USES_RAM is true
SDSORT_CACHE_NAMES is true

The code changes result in file names being pulled from the ALPHA SORT
memory array rather than the SD card.

The code also gives the file count and file index functions their own
variables.  When they shared a common variable the index function
sometimes resulted in the file count being short by 1.

=======================================================================

cardreader.cpp & pins_RAMPS_RE_ARM.h changes

Added another condition to cardreader.cpp to enable getting file names
only from RAM.

pins_RAMPS_RE_ARM.h :
Added comments about the SD card accesses and the LCD display
Combined all versions into this one.

Bob-the-Kuhn committed:
HAL compatible pinsDebug & misc

Re-ARM has been tested.  AVR has not been tested.

1) moved all cpu specific items to files in the low level HAL directory
for that CPU (pinDebug_Re-ARM.h & pinsDebug_AVR_8_bit.h

2) added pinsDebug.h to the top level directory

3) modified HAL_pinsDebug.h to select the correct support file for the
selected CPU

4) Patched sanitycheck to stop throwing false errors.  A long term
solution will be done

5) misc changes & bug fixes
arduino.cpp - included macros.h to fix a missing definition
pinmap_re-arm.h - removed a duplicated line.
pinmapping.h - changed from "ENABLED" to "defined" to fix a compile
error

======================================================================

split SanityCheck up, improve pinsDebug system

======================================================================

switch to latest pins_RAMPS_RE_ARM.h

p3p committed:
Post merge fix


Add macros.h to allow Enabled macro, also need to reference include files directly

p3p committed:
SPI API platform implementation stubs

p3p committed:
Moved CMSIS and other LPC1768 dependencies

Fixes Arduino IDE builds for 8-bit AVR,
misc: Adafruit NeoPixel currently incompatible with Teensy 3.5-6, blacklisted

thinkyhead committed:
Merge pull request #7390 from p3p/bugfix-2.0.x-LPC1768

Bugfix 2.0.x update

p3p committed:
bugfix-2.0.x critical fix (#7401)

* Fix mistake in gitignore file and add in missing core files.

The missing leading slash on "lib" meant all folders names lib in the directory tree are ignored, rather than just the top level PlatformIO lib folder

* Add LiquidCrystal Library and associated headers modified to compile.

Roxy-3D committed:
Fix Large Z correction when nozzle goes off UBL mesh (#7414)

p3p committed:
Revert "SPI API platform implementation stubs" (#7416)

This reverts commit 2dfa6ca72a2e8d43caf4932ca0d35792f0638917.

Revert "Base HAL SPI Changes"

This reverts commit 2afc521b8b6a81b2281a038f1b99a69f4a008e64.

Revert "Initial HAL SPI API"

This reverts commit 58f7ffe09ab5bc034b6510f5204f8d342138abaa.

Bob-the-Kuhn committed:
ultralcd.cpp fix compile error - switch to int16_t

Bob-the-Kuhn committed:
Merge pull request #7417 from Bob-the-Kuhn/2.0.x--ultralcd.cpp-INT-fix

Bugfix_2.0.x - fix ultralcd.cpp compile error (int* to int16_t*)

Commit:
Force legacy "pristine" build dist for TravisCI

TravisCI "trusty" build dist isn't picking up build path causing travis checks to fail.  Reverting to legacy "pristine" dist while we investigate

Bob-the-Kuhn committed:
Merge pull request #7421 from fiveangle/bf_travis_pissed_his_pants

Force legacy "pristine" build dist for TravisCI

Commit:
pristine != precise (#7424)

Commit:
Support travis new "trusty" build environment

move from legacy precise to trusty build image
fix PATH not including buildroot/bin
remove symolic link to ~/bin as trusty travis image doesn't include it in PATH anyway

thinkyhead committed:
Match the license headers in other files

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bugfix-2.0.x

thinkyhead committed:
Merge pull request #7426 from thinkyhead/bf2_license_header

Matching license headers

p3p committed:
Update the delay functions and change the default pinmap for character displays (#7434)

thinkyhead committed:
Merge pull request #7427 from fiveangle/bf2_build_env_to_trusty

Support travis new "trusty" build environment

Bob-the-Kuhn committed:
Set pinMode with every digitalWrite (legacy compatibility)

thinkyhead committed:
Add VSCode to .gitignore

Commit:
try to fix LIN_ADVANCE for 32bit

Tannoo committed:
Re-Arm bugfixes (#7495)

* UBL_correction

* RGB Map function

Bob-the-Kuhn committed:
Servo & PWM support for LPC1768 (#7500)

Bob-the-Kuhn committed:
ST7565

Bob-the-Kuhn committed:
Merge pull request #7502 from Bob-the-Kuhn/Re-ARM-Viki-2

Bugfix_2.0.x - ST7565 printer support (software SPI) - Viki2, Mini-Viki & MaKrPanel #7447 (again)

Roxy-3D committed:
Minor corrections to 32-bit branch (#7514)

Mostly...  clean up some compile time warnings for the 20x4 LCD Display.

Tannoo committed:
LCD_UBL_memory_slot_corrections (#7516)

* LCD_UBL_memory_slot_corrections

Changed the memory slot edit function to work with the
`settings.calc_num_meshes()`

* Add a little more safety margin...

* More corrections

Error handling when the EEPROM is not available.

Bob-the-Kuhn committed:
hybrid PWM system

Uses PWM1 to directly control pins 4, 6 & 11 (servo 0, 1 & 3) and PWM1
generated interrupts to control other pins.

Interupt control of the servo pins had too much jitter so switched all
that we could to PWM1 direct control.  The PWM1 direct control pins have
less than 1 microsecond pulse width jitter while the interrupt
controlled ones can have 20+ microseconds of jitter.

Also added insurance to the servo code in the "disable servo after move"
section.

Bob-the-Kuhn committed:
Merge pull request #7533 from Bob-the-Kuhn/Re-ARM--hybrid-PWM

Servo & PWM support for LPC1768 (hybrid PWM system)

Commit:
add HardwareSerial and SoftwareSerial for Re-ARM.
HardwareSerial has been tested on Uart0 (debug header) and Uart3 (i2c connector)
Software Serial has been tested to work bi-directionally at 9600 and 115200
using pins 6 and 63 on J5, and unidirectionally (write only) at 250000.
The code used to test was Teemuatlut's tmc2208 patch, and a few small changes to main used to echo recieved chars back to a host pc.

thinkyhead committed:
Merge pull request #7546 from kfazz/bugfix-2.0.x-uart-lpc1768

add HardwareSerial and SoftwareSerial for Re-ARM.

thinkyhead committed:
Apply spacing, const to some HAL code

thinkyhead committed:
Other spacing tweaks

thinkyhead committed:
Merge pull request #7555 from thinkyhead/bf2_fix_lin_advance

Try to fix LIN_ADVANCE for 32bit

p3p committed:
Add needed platform defined gaurds to compile with Arduino IDE

Also removed explicit platform HAL directory inclusion from platformio.ini to make sure these errors are caught by Travis in the future

p3p committed:
Fix compile issue when watchdog was disabled

p3p committed:
Temporary fix for odd stepper behaviour with servos

thinkyhead committed:
Merge pull request #7573 from p3p/misc_fixes

Misc fixes

Bob-the-Kuhn committed:
HW SPI

1) Created hardware SPI

2) Modified VIKI & RRD Full Graphics drivers to work with it.

3) Modified HAL.h so that the SPI speeds match the definitions in the SD
card software.

4) disabled software SPI

5) added ENV for J-Link debug probe.

6) made RRG Full graphic run a 1 MHz (independent of SD card rate)
because it had problems at 2 MHz.

===============================================================

kick software SPI up to 250k, 2M & 5M

Bob-the-Kuhn committed:
HW SPI

1) Created hardware SPI

2) Modified VIKI & RRD Full Graphics drivers to work with it.

3) Modified HAL.h so that the SPI speeds match the definitions in the SD
card software.

4) disabled software SPI

5) added ENV for J-Link debug probe.

6) made RRG Full graphic run a 1 MHz (independent of SD card rate)
because it had problems at 2 MHz.

===============================================================

kick software SPI up to 250k, 2M & 5M

add ARM hooks,add soft & hard SPIs to u8g

1) added ARM hooks to u8g library
2) added ARM support routines to u8g
3) added hard & soft SPIs to u8g
4) removed special VIKI routine (no longer needed)
5) updated ultralcd_impl_DOGM.h with correct routine name for VIKI type displays

fix travis problems

added ST7920 support

p3p commented:
It's ok for a stopgap just to enable hardware spi on the Re-ARM but will need a lot of work to generalise, as a rule we can never use things like ` #if ENABLED(IS_PLATFORM)` in Marlin code the HAL layer has to handle it transparently even having to use `#if defined(CPU_32_BIT)` is an indication that something is not abstracted properly yet, The [implementation](https://github.com/MarlinFirmware/Marlin/tree/071c7a78f27078fd4aee9a3ef365fcf5e143531e) I reverted I did plan to finish although probably removing the [logical channels](https://github.com/MarlinFirmware/Marlin/blob/071c7a78f27078fd4aee9a3ef365fcf5e143531e/Marlin/src/HAL/HAL_LPC1768/spi_impl.cpp) (I used them to change mode and frequency depending on the ssel pin) they may have over complicated it a bit, The problem with defining an API is you need to implement it for all current platforms ;)

You seem to have an unrelated commit in this PR?

Bob-the-Kuhn commented:
I'll get rid of the  `#if ENABLED(IS_PLATFORM)` stuff and the extraneous commit.

Long term question - should the top level routines (U8g, SD card, ...) control the slave select lines or should that be part of the low level SPI channel logic?

I can see the channel logic doing it but I don't see a clean path to getting there.  As long as we're still pushing changes from 1.x into 2.0.x we'll always have the danger of breaking 2.0.x.  I'm thinking that the channel concept will have to wait until 2.x has been the main branch for 6-12 months.

I'm thinking that 2.0.x should keep the slave select logic in the top level routines and that the HAL layer should keep the hardware/software SPI hidden from the top level routines.

p3p commented:
I think it may always be necessary to have the slave select controllable through the api, not all hardware has the same timing requirements for it, if you look at my [spi_api.h](https://github.com/MarlinFirmware/Marlin/blob/071c7a78f27078fd4aee9a3ef365fcf5e143531e/Marlin/src/HAL/spi_api.h) I did expose that functionality, I called it slave select override internally I think, as I supported both automatic and api controlled ssel. I was moving towards asynchronous support but never did get that far, that complicates ssel control even more.

Bob-the-Kuhn commented:
How do you like to move forward on this?
- Simple SPI (my code), keep slave select in top level, expect to go to channel logic in 2.1.x
- channel logic now,  keep slave select in top level,  go to channel controlled slave select in 2.1.x
- full channel control now
- ??

p3p commented:
I think moving all spi (and other hardware like i2c) access behind a centralised api is the end goal, and this should be implemented like I started doing so there is global header all platforms use and the implementation in each platform. There are currently far to many places where headers are included by preprocessor conditionals.. 

Having said that we have to get there in baby steps so I think as long as there are no side effects for other platforms your implementation is fine and slave select (through the api by pin) will always be accessible (removing my logical channel logic, just using the hardware channel part, though not sure how to control the frequency (or mode) per ssel without them ), I should have thought of just using #if defined(CPU_32_BIT) to avoid effecting AVR and carried on with what I was doing but I got stressed out with having a half complete api merged upstream. 

have I answered the question? probably not. 1?

Bob-the-Kuhn commented:
OK - move forward with simple right now and wait for you to tell me it's time to take the next step.

thinkyhead commented:
> As long as we're still pushing changes from 1.x into 2.0.x we'll always have the danger of breaking 2.0.x.

That is true. I'll have to be triply careful when rebasing. Hopefully, most of the changes to 1.1.x are at the high level, but I already know the merge resolution will be very meticulous. I've carved out a large block of time over the next few days to do the rebase, followed by the code reorganization.

Obviously I don't want to step on the debugging process. I can post the initial (flat) rebased code as a separate branch so you can try to move any commits following the point in time of this PR onto that. The rebased branch should have no differences in the HAL code from `bugfix-2.0.x` now, and the high-level code will be mostly the same, so it shouldn't be too daunting.

I keep touting the code reorganization, and I'm more excited about getting over that hurdle than any other. But I am holding off because there's a lot of good debug progress being made right now and I don't want to step on it.

That said, there is some good reason to take a break for the code reorganization before continuing in the debug process. Namely, (assuming the AVR version of the HAL is solid) the `2.0` branch can be tagged as the latest current release, with `1.1` in maintenance mode, so we can focus all energy on `2.0.x`. Patches to `1.1.x` will start slowing down while helpful patches to `bugfix-2.0.x` should start to appear from more quarters.

So… my current action item is to post the `bugfix-2.0.x-rebase-31-aug` branch for y'all to examine, and if it's good, replace `bugfix-2.0.x` with it, plus any other patches you want to add, before going straight to the code layout reorganization and as quickly as possible to the `2.0.0` release.

Meanwhile, this PR seems reasonable and can be merged as soon as you're happy with it. I don't mind letting coding standards stuff that triggers my OCD go through, as I'll just do a cleanup pass before release anyway.

p3p commented:
> (assuming the AVR version of the HAL is solid)

I have no idea, I'm not aware of any testers running 2.0 on avr :/

Roxy-3D commented:
> I keep touting the code reorganization, and I'm more excited about getting over that hurdle than any other. But I am holding off because there's a lot of good debug progress being made right now and I don't want to step on it.

Personally...  I don't think this should be a concern.   The amount of code changing as a result of debug is very very small.   We can keep working in the flat (old) branch until the new organization is complete.   And any changes are probably going to be very small and easy to cross over to the new file layout.

I think we should pull the trigger and proceed.

p3p commented:
I agree with Roxy we need the reorg out of the way, most of the debug work is with the LPC1768 HAL only and any changes outside the HAL are being minimised to try and keep the branch sync process as easy as possible. I'm not comfortable saying the AVR platform code is stable enough for a release though it hasn't been worked on (only maintained though changes) since bobc's origional work.

thinkyhead commented:
I agree. Pulling the curtains, locking the doors, putting on the coffee…

thinkyhead commented:
>  I'm not comfortable saying the AVR platform code is stable enough for a release though it hasn't been worked on (only maintained though changes) since bobc's origional work.

I'm comfortable to start testing it as part of the reorg work, and will install it on the 2 machines I have close at hand once it's at that point.

thinkyhead committed:
Clean up some HAL code

thinkyhead committed:
Merge pull request #7596 from thinkyhead/bf2_misc_cleanup_aug31

Clean up some HAL code

thinkyhead commented:
Following up, I have posted the rebased branch for your inspection.

https://github.com/MarlinFirmware/Marlin/tree/rebase-bugfix-2.0.x-aug31

And here are all changes summarized in one commit: https://github.com/thinkyhead/Marlin/commit/bce06c41db6d308228364623fab44613388dd07b?w=1

There were no serious merge conflicts, and everything was easy to resolve. If everything seems at least as stable as the current `bugfix-2.0.x` branch, let me know, and I will move the HEAD of that branch to point at the rebased code. In the meantime, I will also use that branch as the starting-point for the code reorganization, following the same general plan as the `breakup-marlin-idea` branch.

I'll keep the group posted as it develops, and live-tweet about the ordeal if there's anything interesting to report.

Bob-the-Kuhn commented:
I've gotten rid of the `#if ENABLED(IS_PLATFORM)` items.  

I found out that the RRD Full Graphics LCD has problems at 2 MHz and above.  I've changed the RRD routine to run at 1MHz and then put the SPI speed back to the SD card rate.

Bob-the-Kuhn commented:
On eliminating the `#if ENABLED(IS_PLATFORM)` items ...

Re-ARM requires the LCDs to share a SPI with the SD card but the other platforms do not.  The only ways I can see keeping the status quo on the other platforms & keep Re-ARM happy are:
- Separate copies of the Viki & RRD Full Graphics routines.
- Keep the  `#if ENABLED(IS_PLATFORM)` items but disguise it by calling it SHARED_SPI.

I'll commit the SHARED_SPI implementation & see what shakes out.

----

FYI - the Travis error with the Teensy platform appears to be a compiler error.  If I put the line `static void ST7920_SWSPI_SND_8BIT(uint8_t val) {} ` into the file then it's happy.  Gave me a warning that it was unused but it compiled.

Bob-the-Kuhn commented:
I just ran into a complication.  

The 20x4 RepRap Smart Controller, when connected via the Smart LCD adapter, can't use the hardware SPI.  Two of the LCD pins are also the MOSI & SCK pins for the SD card.  When in hardware SPI mode you can't use WRITE nor digitalWrite to control those pins.

We'll probably end up putting together a matrix of controllers and how to interface them to the Re-ARM.


```
LCD                           INTERFACE               SPI mode               
RepRap Full Graphics      |   Smart LCD adapter   |   Either                 
RepRap Smart Controller   |   Smart LCD adapter   |   Software SPI required  
VIKI2                     |   J3 & J5 on Re-ARM   |   Either       
```

I just did another commit to kick the software SPI up to 250KHz.  If we're going to be running software SPI for some of the SD cards then might as well run it at the rate the SD software uses during initialization.

p3p commented:
If it simplifies matters for you just use the ifdefs (for now), you have ended up going down the same path I did, the reason I didn't have time to finish was how complicated it got once you tried to redirect the spi calls from within the lcd files and the consequences of doing that (There was also some reasonable concern from other developers about changing the optimised SPI functions within the lcd files to generic ones). I feel bad having all this code written and you retracing my footsteps ..

p3p commented:
@thinkyhead it seems some header file may have got changed in the new branch causing a lot of errors, I'l see if I can get it to build in a bit.

Bob-the-Kuhn commented:
It's been a fun journey.  Definitely learned a lot.

The LPC1768 just doesn't have enough I/O pins to allow elegant across-the-board LCD interfaces.

I don't see the LPC1768 being used in higher-end printers.  I expect it'll be around for a year or two and then fade into the background.

I'm surprised that there isn't more activity in the Arduino Due compatible motherboards.  More I/O & more hardware SPI channels.  

Once the Re-ARM is fairly useable I'll start playing with the Due.  I have a RAMPS-FD shield so I can run all my steppers & extruders.  It'll replace my AzteegX3Pro which burned up a couple of months ago.

p3p commented:
I think there would have been plenty of IO if they hadn't tried to reserve the ethernet, its more a consequence of the Re-ARM design than the LPC1768 but I still agree with you, I'l probably move back to the STM32F746NG I was using before I started Re-ARM development once its stable, although that may be because I enjoy playing with the hardware more than actually printing things ..

Bob-the-Kuhn commented:
I'm going to explore dumping the hardware SPI all together and just use a software SPI for the Re-ARM board.

Should make things much easier.

My first pass has the SD card running at 4-5MHz.  We won't lose much switching to a software SPI.  Really ugly waveform but that's not a big deal.

p3p commented:
> I'm going to explore dumping the hardware SPI all together and just use a software SPI for the Re-ARM board

hmm someone else did that ;) can't remember who

p3p commented:
@thinkyhead I've isolated the problem to the RAMPS type macros not working with ENABLED/DISABLED, changing the conditions that use them to normal true/false statements seems to fix the build

```
#define IS_RAMPS_EFB (MB(RAMPS_13_EFB) || MB(RAMPS_14_EFB) || MB(RAMPS_14_RE_ARM_EFB) || MB(RAMPS_SMART_EFB) || MB(RAMPS_DUO_EFB) || MB(RAMPS4DUE_EFB))
#define IS_RAMPS_EEB (MB(RAMPS_13_EEB) || MB(RAMPS_14_EEB) || MB(RAMPS_14_RE_ARM_EEB) || MB(RAMPS_SMART_EEB) || MB(RAMPS_DUO_EEB) || MB(RAMPS4DUE_EEB))
#define IS_RAMPS_EFF (MB(RAMPS_13_EFF) || MB(RAMPS_14_EFF) || MB(RAMPS_14_RE_ARM_EFF) || MB(RAMPS_SMART_EFF) || MB(RAMPS_DUO_EFF) || MB(RAMPS4DUE_EFF))
#define IS_RAMPS_EEF (MB(RAMPS_13_EEF) || MB(RAMPS_14_EEF) || MB(RAMPS_14_RE_ARM_EEF) || MB(RAMPS_SMART_EEF) || MB(RAMPS_DUO_EEF) || MB(RAMPS4DUE_EEF))
#define IS_RAMPS_SF  (MB(RAMPS_13_SF)  || MB(RAMPS_14_SF)  || MB(RAMPS_14_RE_ARM_SF)  || MB(RAMPS_SMART_SF)  || MB(RAMPS_DUO_SF)  || MB(RAMPS4DUE_SF) )
```

p3p commented on Marlin/src/HAL/HAL_LPC1768/HAL_spi.cpp:
```diff
@@ -137,28 +124,93 @@
 #else
-  void spiBegin() {
+
+  #include "../../frameworks/CMSIS/LPC1768/include/lpc17xx_pinsel.h"
```
Pretty sure you're not using this now with what you said but the framework headers are actually "system" headers so can be included with just <headerfile.h>, library headers in the platform framework can be include this way too.

thinkyhead commented:
>  I've isolated the problem to the RAMPS type macros

Thanks! I patched it in the "new code layout" branch but forgot to bring it over to the rebase. Should work now!

```cpp
#if   MB(RAMPS_13_EFB) || MB(RAMPS_14_EFB) || MB(RAMPS_14_RE_ARM_EFB) || MB(RAMPS_SMART_EFB) || MB(RAMPS_DUO_EFB) || MB(RAMPS4DUE_EFB)
  #define IS_RAMPS_EFB
#elif MB(RAMPS_13_EEB) || MB(RAMPS_14_EEB) || MB(RAMPS_14_RE_ARM_EEB) || MB(RAMPS_SMART_EEB) || MB(RAMPS_DUO_EEB) || MB(RAMPS4DUE_EEB)
  #define IS_RAMPS_EEB
#elif MB(RAMPS_13_EFF) || MB(RAMPS_14_EFF) || MB(RAMPS_14_RE_ARM_EFF) || MB(RAMPS_SMART_EFF) || MB(RAMPS_DUO_EFF) || MB(RAMPS4DUE_EFF)
  #define IS_RAMPS_EFF
#elif MB(RAMPS_13_EEF) || MB(RAMPS_14_EEF) || MB(RAMPS_14_RE_ARM_EEF) || MB(RAMPS_SMART_EEF) || MB(RAMPS_DUO_EEF) || MB(RAMPS4DUE_EEF)
  #define IS_RAMPS_EEF
#elif MB(RAMPS_13_SF)  || MB(RAMPS_14_SF)  || MB(RAMPS_14_RE_ARM_SF)  || MB(RAMPS_SMART_SF)  || MB(RAMPS_DUO_SF)  || MB(RAMPS4DUE_SF)
  #define IS_RAMPS_SF
#endif
```
The `bf2_new_code_layout` branch is coming along well, btw. I have all the files distributed to sub-folders, but I still have to break up the G-codes. I will aim to be done with that tonight or tomorrow and have a fully working new hierarchy that we can develop from.

p3p commented:
Looks like it's going well, only had a quick look but to build with Arduino IDE the frameworks folder can't be moved inside src.

Bob-the-Kuhn commented:
I didn't drop the hardware SPI, just made it inactive by default.

The software SPI runs at rates of 5M peak, 2M peak and 250K peak.  Wave forms are really ugly.

I also included a link to Chris' more sophisticated hardware SPI incase LPC1768 moves to interrupt driven SPI transfers.

Bob-the-Kuhn commented:
@p3p - I think this is about as far as we can go without doing a total re-write of all the SPI uses within Marlin.

Please merge this if you're happy with it.

p3p commented:
I'l look it over and test it but I think we should wait for @thinkyhead to finish his reorg endeavour before merging anything new, .. although I'm not trusted with that power anyway ;)

tcm0116 commented:
>The 20x4 RepRap Smart Controller, when connected via the Smart LCD adapter, can't use the hardware SPI. 

Don't know if you say my comment in the other thread about how the pins are mapped (https://github.com/MarlinFirmware/Marlin/issues/7076#issuecomment-326715329), but I think this can be resolved by moving all of the control signals for the RRD Smart Controller to the Ethernet connector on the Re-ARM board. This should also fix the issue of data corruption on the Smart Controller LCD that currently exists with the software SPI.

The Full Graphic Controller should still be usable via the adapter and the split cable that is provided with the Re-ARM board.

Bob-the-Kuhn commented:
Yes, I watched the other thread.

What Re-ARM needs is a hardware adapter that spans the Re-ARM connectors and provides signals to all the pins on the two 10 pin connectors.

victorpv commented:
In my opinion going to a 32bit fast mcu and then keeping it busy doing software SPI seems counterproductive. 
Not sure all the barriers to doing hardware SPI due to pins conflicts, but wouldn't be easier to move some LCD signals to other GPIO pins, even if that requires a PCB adapter?
or just stop support for some types of displays if they conflict with hardware SPI?
If you limit the hardware at this point due to compatibility with some other hardware, I think it will be much harder to remove those limitations later. 

In turn from a user point of view if when you are buying a Re-Arm you know that you have to use X, Y or Z display boards, or a small PCB adapter, then you will likely just do so. I think the chances of someone getting upset because the Re-Arm performance is not what they though, because of all the problems derived from software SPI or any other limitation, are larger than someone being upset at having to buy a new display or a small adapter if they are told so from the start.
I understand the idea on trying to support as much legacy hardware as possible, but seems like that's going to handycap the potential performance of the board.
One other option, would it be possible to have Software and Hardware SPI be used automatically depending on the display etc being used? Keep that abstracted behind the HAL, so no extra configurations on configuration.h, but have the HAL decide what SPI to use depending on what's configured.

tcm0116 commented:
>What Re-ARM needs is a hardware adapter that spans the Re-ARM connectors and provides signals to all the pins on the two 10 pin connectors.

I've been working on a set of ribbon cables to perform this function. If
we're willing to re-map some of the pins when using the smart controller,
we can get everything on J3 and J12. I'll provide the details later when I
get it a little further along.

tcm0116 commented:
@Bob-the-Kuhn Below is the table I've come up with that re-maps the pins for the RepRap Discount Smart Controller to J3 and J12 on the Re-ARM board. By building ribbon cables in this configuration for that display, we should be able to use the hardware SPI to access the SD card.

UPDATE: I had VCC and GND switched in the table, so I've made a correction.

Signal | EXP Pin | Re-ARM Pin | Pin
-- | -- | -- | --
VCC | EXP1-10 | J12-14 |  
GND | EXP1-9 | J12-2 |  
LCD7 | EXP1-8 | J12-8 | 75
LCD6 | EXP1-7 | J12-6 | 73
LCD5 | EXP1-6 | J12-4 | 71
LCD4 | EXP1-5 | J12-3 | 70
LCDRS | EXP1-4 | J12-5 | 72
LCDE | EXP1-3 | J12-7 | 74
BTN ENC | EXP1-2 | J12-10 | 77
Beeper | EXP1-1 | J12-11 | 78
PB3 (MISO) | EXP2-1 | J3-6 | 50
PB1 (SCK) | EXP2-2 | J3-9 | 52
BTN EN2 | EXP2-3 | J3-2 | 31
SD CSEL | EXP2-4 | J3-5 | 53
BTN EN1 | EXP2-5 | J3-4 | 33
PB2 (MOSI) | EXP2-6 | J3-10 | 51
SD DET | EXP2-7 | J3-1 | 49
RESET | EXP2-8 | J3-7 | 16
GND | EXP2-9 | J3-11 |  
KILL | EXP2-10 |   |  

>One other option, would it be possible to have Software and Hardware SPI be used automatically depending on the display etc being used?

I'm not sure this is necessary because the 20x4 LCDs will still have the corruption issue unless at least LCDE is moved to J12 along with the LCD5-7 lines. At that point, you might as well also move LCD4, which will allow for hardware SPI.

In order for the 20x4 LCDs to work correctly and without corruption, LCDE and LCD4-7 need to be moved to J12 on the Re-ARM. It would be possible to keep the EXP2 connector running to the RAMPS adapter and only route EXP1 to J12 on the Re-ARM, but if you're going to be re-terminating one entire cable, why not just do both and ditch the adapter? However, if we want to use that approach, then the only change to the table above is that the RESET pin would be re-mapped to pin 41.

Bob-the-Kuhn commented:
I hope you've been using this [wiki](http://reprap.org/wiki/Re-ARM) when developing your tables.  Should save a bunch of time.

Don't worry about softSPI vs. hardSPI.  I doubt that you'll see a difference in performance between the two.   You can definitely see a difference using a logic analyzer but not if using the LCD.

The only reason to use one vs. the other is if the LCD shares MOSI or SCK with the SD card.  In that case the softSPI is required.  If a nice cable/interface comes into existence then it won't matter.

tcm0116 commented:
>NOTE - The mega2560 pinout used two SPI channels but the Re-ARM only routes one to this adapter. The result is garbage/lines appearing on the RepRap Discount Full Graphic display when the SD card is accessed. Sometimes the garbage/lines remains on the "Print from SD card" submenu screen after the SD card accesses are finished.

I don't believe this statement from the wiki is true. The Arduino Mega only has one SPI channel. There are independent chip select signals running to the LCD and SD card in the full graphic controller, so only one SPI channel is required so long as the CS signals are set properly when reading/writing to either device. 

For the standard smart controller, some of the signals for the 4-bit parallel interface to the LCD controller use the same pins as the SPI signals used by the SD card on the Re-ARM board. My table above attempts to address this issue by moving those shared signals to J15 on the Re-ARM, in addition to going ahead and moving all of the signals to J3 and J15 to eliminate the need for the RAMPS adapter. 

As I previously indicated there are several solutions to the shared signals issue for the standard smart controller, with the bare minimum being to move LCDE and LCD4-7 to J15 on the Re-ARM board. 

Regarding the corruption issue with the full graphic controller, there must be an issue with the HAL as the chip select signals for the LCD and SD card are independent pins on the Re-ARM board, so they can safely share a single SPI port.

Bob-the-Kuhn commented:
On the 2560, for the RRD Full Graphic, there is one SPI channel for the SD card and a software SPI for the LCD.

The RRD Full Graphic LCD controller responds to the MOSI line even when the slave select is inactive.  That's not a problem on the 2560 but on the Re-ARM the SCK & MOSI lines are shared between the LCD and the SD card.  The result on the Re-ARM is varying degrees of garbage on the screen.  To get around this we use the SD card cache option to minimize SD card accesses and we changed the software so that the "print from SD card" directory listing is drawn only after the SD card accesses are completed.

If you do a M20 from the host you can see momentary garbage on the LCD in the Re-ARM system.

victorpv commented:
@Bob-the-Kuhn  the RRD Full graphic display uses software SPI? I thought it used a parallel interface and that's why it reacted to sck and mosi (cause they mapped to LCD_RS or LCD_E. If it uses SPI and acts when CS is down, that's some design defect in the display isn't it?

Bob-the-Kuhn commented:
Yes, the RRD Full graphic display uses a software SPI.

Yes, reacting to SCK & MOSI when slave select is inactive is unusual..  The data sheet says something like "undefined results" when that happens.

tcm0116 commented:
It looks like the full graphic controller uses a ST7920 LCD controller/driver, which states the following:

>When connecting several ST7920, chip select（CS）must be used. Only when（CS）is high the serial clock（SCLK）can be accepted. On the other hand, when chip select（CS）is low ST7920 serial counter and data will be reset. Transmission will be terminated and data will be cleared. Serial transfer counter is set to the first bit. For a minimal system with only one ST7920 and one MPU, only SCLK and SID pins are necessary. CS pin should pull to high.

As such, de-asserting the chip select should perform the expected function. The next paragraph in the data sheet states:

>ST7920’s serial clock（SCLK）is asynchronous to the internal clock and is generated by MPU. When multiple instruction/data is transferred instruction execution time must be considered. Must wait for the previous instruction to finish before sending the next. ST7920 has no internal instruction buffer area.

I wonder if there's a timing difference between the 8-bit AVR processor and the LPC1768 such that the commands are being sent too fast to the LCD controller, causing days corruption. According to the Instruction Set tables, it looks like the standard command execution time is 72 us.

bobc commented:
> It looks like the full graphic controller

Looks like, or *is* using? These LCD controllers have many clones, it's
quite possible the sort of cheap clone the Chinese typically use does not
work to the original spec.

The problem was discovered by the Smoothieware team and is a known problem,
at least for them.

On 7 September 2017 at 01:32, Thomas Moore <notifications@github.com> wrote:

> It looks like the full graphic controller uses a ST7920 LCD
> controller/driver, which states the following:
>
> When connecting several ST7920, chip select（CS）must be used. Only
> when（CS）is high the serial clock（SCLK）can be accepted. On the other hand,
> when chip select（CS）is low ST7920 serial counter and data will be reset.
> Transmission will be terminated and data will be cleared. Serial transfer
> counter is set to the first bit. For a minimal system with only one ST7920
> and one MPU, only SCLK and SID pins are necessary. CS pin should pull to
> high.
>
> As such, de-asserting the chip select should perform the expected
> function. The next paragraph in the data sheet states:
>
> ST7920’s serial clock（SCLK）is asynchronous to the internal clock and is
> generated by MPU. When multiple instruction/data is transferred instruction
> execution time must be considered. Must wait for the previous instruction
> to finish before sending the next. ST7920 has no internal instruction
> buffer area.
>
> I wonder if there's a timing difference between the 8-bit AVR processor
> and the LPC1768 such that the commands are being sent too fast to the LCD
> controller, causing days corruption. According to the Instruction Set
> tables, it looks like the standard command execution time is 72 us.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7587#issuecomment-327648749>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA7VRxH4JOsbwokY5XclHe0JNQljC-wmks5sfzmEgaJpZM4PINLr>
> .
>

p3p commented:
I spent a long time staring at signal traces while originally getting the RRGLCD controller working with the lpc1768 because of the sd card corruption and after a lot of experimenting (and reading the datasheet ;) ) the only conclusion that made sense was sending sck when cs is low is still interpreted by the controller, (there was a long thread in one of the many and long 2.0 threads about it)

bobc commented:
I found this old post : https://github.com/olikraus/u8glib/issues/221 which
goes back to 2015. I think if 3 people found the problem independently,
there probably really is a problem with the LCD controller.

On 7 September 2017 at 01:54, Chris Pepper <notifications@github.com> wrote:

> I spent a long time staring at signal traces while originally getting the
> RRGLCD controller working with the lpc1768 because of the sd card
> corruption and after a lot of experimenting (and reading the datasheet ;) )
> the only conclusion that made sense was sending sck when cs is low is still
> interpreted by the controller, (there was a long thread in one of the many
> and long 2.0 threads about it)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/7587#issuecomment-327651693>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA7VR-a74RRj5-RJQmc_gUkZfZL3WMXYks5sfz7CgaJpZM4PINLr>
> .
>

p3p commented:
> Serial mode: Chip select.
> 1: chip enabled;
> 0: chip disabled.
> When chip is disabled, SID and SCLK
> should be set as “H” or “L”. Transcient
> of SID and SCLK is not allowed. 

That's a quote for the st7920 data sheet, it makes absolutely no sense and completely invalidates the purpose of a CS line, unless I'm misunderstanding it (which is likely)

tcm0116 commented:
>I think if 3 people found the problem independently,
there probably really is a problem with the LCD controller.

That's both enlightening and disappointing at the same time, but I would say that puts the final nail in the coffin. 

>the RRD Full graphic display uses a software SPI.

Am I the only one who thinks this is backwards since the LCD is accessed continuously as opposed to the SD card, which is accessed in spurts (if even at all)?

Since the original RRD Smart Controller used a parallel interface, it makes sense that the hardware SPI was used for the SD card. 

Perhaps the solution for both controllers is just to move directly to J3 and J12 on the Re-ARM board, that way we can have hardware SPI for one device and software SPI for the other.

victorpv commented:
I didn't mean to doubt there was an issue with the controller, just try to understand it since I thought it was a parallel interface.
Is there any chance of using MOSI1, SCK1 (p0.9, p0.7) or those are already used for something else?

tcm0116 commented:
>I didn't mean to doubt there was an issue with the controller, just try to understand it since I thought it was a parallel interface.

Me neither. This has been an interesting learning experience. FYSA, the RRD Smart Controller is a 4-bit parallel interface, and the RRD Full Graphic Controller is a serial interface (although it can operate in a parallel mode, but those signals after not brought out on the board). 

I don't see p0.7 or p0.9 on the Re-ARM schematic (http://panucattdevices.freshdesk.com/helpdesk/attachments/1047536701). 

My table above is for the Smart Controller, but I think it would work for the Full Graphic controller if would be possible to use both hardware and software SPI as in the 2560. @Bob-the-Kuhn - do you think that would be difficult to get working?

victorpv commented:
Found where those are from the pinmap file :(
```cpp
//  {0,6},        //  DIO80  SD-CS            on board SD card
//  {0,7},        //  DIO81  SD-SCK           on board SD card
//  {0,8},        //  DIO82  SD-MISO          on board SD card
//  {0,9},        //  DIO83  SD-MOSI          on board SD card
```
Seems like there are only 2 spi ports and both are used.

tcm0116 commented:
I wonder if it would be possible to print from the onboard SD card...

Roxy-3D commented:
> I wonder if it would be possible to print from the onboard SD card...

I'm sure it would be.  But we would end up with two SD-Card object.   On the 32-bit processors, that probably isn't a big deal.  But we would still have to re-work the code to let the user pick which SD card was going to supply the file.    

The SD Card on the 32-bit board is providing EEPROM emulation.   I'm sure this can be debated, but it might be good to just leave in in the socket and never unplug it.   Because if that card disappears, all the saved data disappears.

My vote from and 'ease of use' perspective would be we leave that one always installed and don't touch it.   And the other socket (on the LCD Panel) is where transient GCode files get presented to the firmware.

Differing view points and debate are welcome!!!

victorpv commented:
I understood @tcm0116 suggesting that we use the onboard sdcard for printing, so the other SPI port is used for the display rather than for a second sd card. So there would not be 2 sd card objects, just the onboard one. That may be confusing.
I think that using HW SPI for the display LCD, and software SPI in separate pins for the LCD makes the most sense. It will just need to have different pins routed to the display so they are not shared.

tcm0116 commented:
I wired up my RRD Smart Controller per the table I provided above, and am able to have full control of the controller. I used the following in the pins_RAMPS_RE_ARM.h file:

```
  #ifdef ULTIPANEL   
    #define LCD_PINS_D4         70  // (ENET_MOC) J12-3
    #define LCD_PINS_D5         71  // (ENET_MDIO) J12-4
    #define LCD_PINS_D6         73  // (ENET_RX_ER) J12-6
    #define LCD_PINS_D7         75  // (ENET_RXD1) J12-8
    #define LCD_PINS_RS         72  // (REF_CLK) J12-5
    #define LCD_PINS_ENABLE     74  // (ENET_RXD0) J12-7
    #define BTN_ENC             77  // (ENET_TX_EN) J12-10
    #define BEEPER_PIN          78  // (ENET_TXD0) J12-11
    #define KILL_PIN            16  // J3-7 & AUX-4
  #else
    #define LCD_PINS_D4         52  // (SCK)  J3-9 & AUX-3
    #define LCD_PINS_RS         16  // J3-7 & AUX-4
    #define LCD_PINS_ENABLE     51  // (MOSI) J3-10 & AUX-3
    #define BTN_ENC             35  // J3-3 & AUX-4
    #define BEEPER_PIN          37  // not 5V tolerant
    #define KILL_PIN            41  // J5-4 & AUX-4
  #endif
```

My next step is to work with the changes in this PR to see if I can get hardware SPI working for the SD card and software SPI working with the RRD Full Graphic controller, which should solve the LCD corruption issue.

tcm0116 commented:
Here's picture so you can see how I have it wired up:

![img_20170907_083117](https://user-images.githubusercontent.com/1868773/30165855-2c6e89b2-93a7-11e7-8990-cc21ea0a9a08.jpg)

Bob-the-Kuhn commented:
I'm back.  Did a clean install of Win10.  I've got about 80% of the functionality back.  Still can't get platformio running properly.

----

There's multiple hardware & software SPIs,

It really doesn't matter if a hardware or software SPI is used to run the LCD and/or the SD card as long as they're running in the MHz range.  The LCD is updated once a second with the update lasting 1-5 mS.

The LPC1768 has two hardware SPIs.  These can run much faster than the software SPIs.  One is dedicated to the onboard SD card.  The other is used for the SD controller on the LCD display and, optionally, can be used to run the LCD.

The U8g software for SPI based LCD controllers can use the U8g software SPI or the SD card's SPI.  This is selected by the routine specified in the **ultralcd_impl_DOGM.h** file.  You can tell the difference by whether the MOSI & SCK pins are enumerated in the arguments.  If they are present then the U8g software SPI is used.

The way my proposed software is written, the SD card SPI can be either the LPC's hardware SPI or a software SPI.  The software SPI is enabled by default.  This is because it allows for more flexibility.  If the LPC's (hardware) SPI module is driving MOSI & SCK then WRITE(MOSI, x) and WRITE(SCK, x) are ignored.  This means that the u8g software SPI can't use those lines.  That's the case on the RRD Full Graphic when using the standard (Smart) adapter.

The way I like to refer to the SPIs are:
**SD card SPI** - can be hardware or software based
**LCD SPI** - can be either the U8g software SPI or can share the SD card SPI

----

Things sure get complicated when there are shared pins.

In my opinion we're stuck with using the RRD Full Graphic via the standard (Smart) adapter just because that's what the average joe will be using.  

The only way I see that changing is if a new adapter PCB becomes readily available.  That won't happen until after the 2.0.x software is released for general use so we need to keep the software flexible enough that either adapter can be used.

tcm0116 commented:
>Still can't get platformio running properly.

Even at the command line?

>the SD card SPI can be either the LPC's hardware SPI or a software SPI.

What about the option for using hardware SPI for the **SD card SPI** and software SPI for the **LCD SPI**, which is how it is done on the 2560? I'll provide some further detail below.

> In my opinion we're stuck with using the RRD Full Graphic via the standard (Smart) adapter just because that's what the average joe will be using.
> 
> The only way I see that changing is if a new adapter PCB becomes readily available.

We already know that we have to move some of the wires for the RRD Smart Controller because they simply weren't included in the Re-ARM. We also know that the RRD Full Graphic Controller has a design flaw such that trying to utilize a shared SPI channel results in garbage being displayed on the LCD when trying to access the SD card.

As such, it seems to me that it would be in our best interest to go ahead and define what we think the best solution is. This way, we can get ahead of the curve, and whoever makes a Re-ARM to RRD Smart or Full Graphic Controller adapter or wiring harness will hopefully do what we want instead of us being at the whim of someone else.

Perhaps we can provide a configuration option to use independent SPI channels for the RRD Full Graphic Controller, which would then require that the controller be connected to J3 and J12 of the Re-ARM per a table similar to mine above.

p3p commented:
@tcm0116 We have to support (as best we can) what an average user will do when upgrading to a Re-ARM and that is just plug everything in (hopefully using the cable that is provided with the Re-ARM) apart from the 2 lines of corruption that appear and disappear straight after sd card accesses (RRD FGLCD) mine is working flawlessly in this configuration, the 20x4 displays do require a cable modification so there is nothing we can do in software to work around that. Not saying we aren't going to allow arbitrary software and hardware spi connections (I was working on that before bob) but its not going to be the default setup.

Bob-the-Kuhn commented:
I agree on the 2560.  The hardware SPI is used for the SD card and the U8g software SPI is used for the LCD controller.

Bob-the-Kuhn commented:
I've gotten the U8g generic software SPI working.  Without interrupts the SPI runs at the expected 250kHz but ... when the USB port is active it slows down to about 100kHz.  Even at that speed I don't see a noticeable lag on the LCD screen when playing with the encoder.

I did notice that the name of the Viki2 driver is different in the UGlib-ARM library than in the Marlin 1.x U8g library.  I'll see if there's any other driver name changes that need to be done.

After that I'll start working on the hardware U8G SPI support.  And then 4 bit parallel,. And then ??

----

I was really surprised to see the USB interrupt taking up about 45% of the CPU's time with the controller idling (USB plugged into my PC, host software NOT running and no printing/movement).  It kicks off every 10uS and runs for 4.5uS.  The LCD updates only happen when the USB interrupt is inactive.

Definitely glad that the USB interrupt is low priority.  It does make me wonder if the USB overhead could suck up enough time during a real print that the planner would be the limiting factor.  Definitely something to look at when doing system tuning .

tcm0116 commented:
@p3p understood. I'm just thinking of what will happen when people start to
install these 32-bit boards in their printers and then we start to get a
flood of issues because the 20x4 LCD isn't working or there's corruption on
the full graphic display.

If we're prepared with solutions (even if they require a custom cable,
adapter, or whatever that isn't yet produced) that are documented in a wiki
and have options that can be enabled in the configuration, then the
transition to 32-bit will be a lot less painful for the average user.

Part of that solution could be some smarts in the software that will choose
the appropriate method of SPI communications. For example, if the SD card
and full graphic controller are both ok the same SPI bus (the default),
then we know we have to use software SPI. However, if the SD card is on the
hardware SPI pins and the LCD is on different pins, then we can use
hard-earned SPI for the SD card and software SPI for the LCD. Ideally, we
provide an compilation option that, when disabled, uses the default pins,
and when enabled, uses a different set of pins that don't conflict. Those
pins would have to be board-specific, so some work would have to go into
providing default values for these alternate pins. We could defer that part
for now, and just indicate that alternate pins will need to be defined in
order to resolve the corruption issue on the full graphic display.

For the 20x4 LCDs, we could have a configuration option that is disabled by
default that basically causes an error to be thrown on compilation if the
RRD Smart Controller is enabled. As part of the description of this "guard"
option, we can document how the controller needs to be wired, and that
enabling the option will clear the compilation error. This would at least
help ensure that someone building the firmware would become aware that
there is an issue that have to address.

I'm more than happy to help look into things like this that can make life
easier for the users when we release 2.0 into the wild. Just let me know.

p3p commented:
> I was really surprised to see the USB interrupt taking up about 45% of the CPU's time with the controller idling

@Bob-the-Kuhn It shouldn't be running that often (USB2 has a 1ms polling interval), that's worrying almost like the usb stack isn't resetting the interrupt flag? I'l look into it

p3p commented:
@Bob-the-Kuhn can you enable the SOF event, https://github.com/MarlinFirmware/Marlin/blob/rebase-bugfix-2.0.x-aug31/Marlin/frameworks/CMSIS/LPC1768/lib/usb/usbcfg.h#L128 and see if that changes it, it appears to disable the interrupt reset if its not enabled, but its a lot of code to understand.

thinkyhead commented:
> leave [the onboard SD card] always installed and don't touch it

Seems prudent! Maybe if a "remove" signal is detected, Marlin should go to KILL.

p3p commented:
@thinkyhead Unfortunately we don't have a card detect signal on that socket, best we can do is detect a timeout on SPI comms. I'm not sure it is a failsafe situation though, a warning message about being unable to save config (EEPROM) data perhaps, and if the firmware is setup to use EEPROM but cant access it on boot display a warning about that too. Filesystems are another thing that we will need to centralise at some point if we want to have a general solution outside the HAL for multiple cards, chanfs does support multiple logical drives, but obviously the SD Card code is another part of Marlin that is highly optimised for AVR thats why I just threw in another FAT library for Re-ARM to use internally.

thinkyhead commented:
@p3p I'm all for using platform-specific libraries where it matters for hardware access. And we can move any features that aren't platform-agnostic into the HALs also.

At the moment I'm refactoring the movement and leveling code, and _should_ have a passing build to post pretty soon.

Bob-the-Kuhn committed:
add ARM hooks,add soft & hard SPIs to u8g

1) added ARM hooks to u8g library
2) added ARM support routines to u8g
3) added hard & soft SPIs to u8g
4) removed special VIKI routine (no longer needed)
5) updated ultralcd_impl_DOGM.h with correct routine name for VIKI type displays

Bob-the-Kuhn commented:
I've just pushed a snapshot of my work on the LCDs. Feedback is encouraged.

I've been able to hook into the u8g library and add routines.  So far I've implemented general solutions for the hardware SPI and the software SPI.

The implementation is far from elegant.  The big problem is getting the "#include" statements correct.  One example is the u8g hardware SPI.  Ideally I'd include the LPC1768 in it so I could access spiSend.  I ended up including the u8g hardware SPI at the end of the LPC1768 SPI code.

The u8g software SPI is running at about 800KHz peak.  The hardware SPI is running about 5MHz peak.

----

1) added ARM hooks to u8g library
2) added ARM support routines to u8g
3) added hard & soft SPIs to u8g
4) removed special VIKI routine (no longer needed)
5) updated ultralcd_impl_DOGM.h with correct routine name for VIKI type displays

Bob-the-Kuhn committed:
fix travis problems

p3p commented:
@Bob-the-Kuhn Did you check whether my change above helped with the high USB interrupt load?

Bob-the-Kuhn commented:
I did not check.  I'll play with that tonight or tomorrow.

Bob-the-Kuhn committed:
added ST7920 support

Bob-the-Kuhn committed:
Merge branch 'Re-ARM-HW-SPI-with-VIKI-&-RRD-Full-graphics-support' of https://github.com/Bob-the-Kuhn/RCBugfix-M43-and-spindle into Re-ARM-HW-SPI-with-VIKI-&-RRD-Full-graphics-support

Bob-the-Kuhn commented:
I just pushed a commit that added support for the st7920. 

That should allow Re-ARM to handle all the dogm displays except for the I2C ones.

Since the 2004 Smart controller works on Re-ARM that means the liquid crystal displays will also work.

@p3p - curiosity question - why do the liquid crystal displays work?  I assume there is a mechanism where the library knows that pinMode and digitalWrite are the correct routines to use.  If my memory is correct there was something you did to enable the 2004 for Roxy.

----

I just tried to squash the commits on this branch but apparently failed.  The new Desktop app for Windows doesn't give me the powershell option.  Wish it was still around.

p3p commented:
@Bob-the-Kuhn LiquidCrystal is a pretty simple library, and only uses the standard Arduino API functions, so the way we implement the Arduino functions in the HAL platforms that arn't Arduino based makes the library work without any problems, I just copied the files into the framework and changed a include statement or two.

fiveangle commented:
I wonder if it would be possible to print from the onboard SD card...> I'm sure it would be. But we would end up with two SD-Card object. On the 32-bit processors, that probably isn't a big deal. But we would still have to re-work the code to let the user pick which SD card was going to supply the file.

I made selection of which SD card to use a [compile-time option in the PRINTRBOARD_REVF](https://github.com/MarlinFirmware/Marlin/blob/1.1.x/Marlin/pins_PRINTRBOARD_REVF.h#L171-L180) pins for AVR.  I think supporting both SD cards at runtime is gratuitous.

> My vote from and 'ease of use' perspective would be we leave that [onboard sd card] always installed and don't touch it. And the other socket (on the LCD Panel) is where transient GCode files get presented to the firmware.

I see your point.  Yeah, it's probably best to leave the onboard dedicated for EEPROM emulation.

Bob-the-Kuhn commented:
I've done some looking at the USB interrupt issue.

I enabled the SOF event by setting the bit to 1 in the config file.  Didn't affect anything.

It looks like the problem is the EPxTX bit.  In the USBEpIntSt register  I see 0x800 (with an occasional 0x820) before the mapped drive shows up and afterwards I see 0x20.  That's the values at the beginning of the ISR.  At the end USBEpIntSt is 0 but 5.5uS later the register is again reading 0x800 or 0x20.

Here's the modified ISR.  I use pin 71 & a logic analyzer to get the times.  I added a routine to the M43 command to print on demand the start and end values.
[usbhw.cpp.txt](https://github.com/MarlinFirmware/Marlin/files/1290888/usbhw.cpp.txt)

Bob-the-Kuhn commented:
If I disable the NAK interrupt & enable the SOF interrupt I see just the 1mS spaced ISRs.  Unfortunately if I then try to connect via the host software, the controller locks up.

Bob-the-Kuhn commented:
These changes are being implemented in PR #7746
